### PR TITLE
fix(component): sane defaults for Nginx updateStrategy and autoscaling

### LINKED ISSUE #855: HPA config for Nginx platform controller and default backend conflicting with RollingUpdate
For both the platform controller and default backend deployments the default replica count is 1, whereas the upgrade strategy is as follows:
``` 
replicas: 1
...
strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 25%
      maxSurge: 25%
```

What myself and @j-zimnowoda  witnessed is that as an update is initiated, new pods are created in an attempt to do the rolling update, but are instantly terminated, likely due to the HPA scaling it down but I am researching this.

From below:
possible solution is to set a maxUnavailable of 50% and maxSurge of 100% to allow the RollingUpdate to override the HPA during the update, which would support situations in which the HPA specifies 1 replica.

In the longer term we can implement validation that will prevent such configs being applied to the cluster.

### PR DESCRIPTION
**Pull request by staticvoid255**
fix #855

### CODE REVIEW

**Review by j-zimnowoda — COMMENTED** [review]
Could you provide some numbers?
Let's say autoscaler is set from 1 to 100
in case of rolling update what number of Pods can we expect? lets say for 1, 10, 50, 100 replicas

File: values/ingress-nginx/ingress-nginx.gotmpl [hunk 1]
```diff
@@ -153,6 +158,13 @@ defaultBackend:
     requests:
       cpu: 10m
       memory: 10Mi
+  autoscaling:
+    enabled: {{ $n.autoscaling.enabled }}
+    minReplicas: 1
```
[the remark this answers is not in the document]
Comment (staticvoid255) [reply values/ingress-nginx/ingress-nginx.gotmpl:163]: good idea!

**Comment by staticvoid255** [comment]
> Could you provide some numbers? Let's say autoscaler is set from 1 to 100 in case of rolling update what number of Pods can we expect? lets say for 1, 10, 50, 100 replicas

The behaviour is really complex. For more details see https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#autoscaling-during-rolling-update
It's tough to know what the actual behaviour is at scale I'm afraid, but at small counts it should do this:

Let's say you have a minReplicas of 1 and maxReplicas of 2, with current proposed maxUnavailable (50%) and maxSurge(100%). We will refer to each replica as R1 and R2, starting out with 1 replica:

1. Update begins
2. R1 is requested to terminate but can't do so until R2 is available
3. R2 is created in conformance with maxReplicas (2), and maxSurge (100% surge means 2 replicas at current replica count of 1)
4. R2 becomes available
5. R1 is terminated
6. We are back at minReplicas

In this case we always satisfy the constraints and have nothing to worry about. What I am in fact worried about and will have to test is if maxReplicas in the HPA constitutes the _absolute total_ max number **including** maxSurge, for example:

1.  if you are already scaled out to 2 replicas (the maximum), _will_ maxSurge of 100%  allow **up to** 4 replicas to be created given 2 is the current _actual_ replica count??  This is preferrable.
2. ..or will it just start terminating all surge pods above the maxReplica count in the config? (boo very bad booo)

**If the 1st case (good):**
Just because it **can** create up to 4 replicas, will it actually do so?
I don't think so (which is good), as at high scale it can just start swapping in replicas 1-by-1 or whatever basis it wants as long as at least 50% are available.

**If the 2nd case (bad)
This will get very complex, as we must then always make sure that the maxSurge count does not exceed the maxReplica count, while still making sure that the are enough free replicas to satisfy the maxUnavailable requirement and perform the update. In this case a lower maxUnavailable is preferrable, but doesn't solve the issue at our current  small scale.

**Comment by staticvoid255** [comment]
There's also the simple alternative of specifying a minReplica count of 2, and a maxReplica count of 4, in which case the existing maxUnavailable of 25% can always be satisfied without requiring any surge.

**Pull request closed without merging**